### PR TITLE
feature(wyh): add PLR algorithm

### PR DESCRIPTION
**Pull request by Weiyuhong-1998**
## Description
PLR algorithm, including level replay, new pipeline

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/worker/level_replay/level_sampler.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+            staleness_weights = staleness_weights * (1-self.unseen_seed_weights)
+            z = np.sum(staleness_weights)
+            if z > 0: 
+                staleness_weights /= z
+
+            weights = (1 - self.staleness_coef)*weights + self.staleness_coef*staleness_weights
+
+        return weights
+
+    def _score_transform(self, transform, temperature, scores):
```
Comment (PaParaZz1) [inline ding/worker/level_replay/level_sampler.py:286]: only rank

File: ding/worker/level_replay/level_sampler.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
+from collections import namedtuple
+import numpy as np
+import torch
+
+class LevelSampler():
+    def __init__(
```
Comment (PaParaZz1) [inline ding/worker/level_replay/level_sampler.py:6]: add default config

File: ding/worker/level_replay/level_sampler.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
+from collections import namedtuple
```
Comment (PaParaZz1) [inline ding/worker/level_replay/level_sampler.py:1]: move this file to `ding/data/level_sampler.py`

File: ding/worker/level_replay/level_sampler.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+import numpy as np
+import torch
+
+class LevelSampler():
```
Comment (PaParaZz1) [inline ding/worker/level_replay/level_sampler.py:5]: add paper link in overview comment

File: ding/worker/level_replay/level_sampler.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        self.partial_seed_scores = np.zeros((num_actors, len(seeds)), dtype=np.float)
+        self.partial_seed_steps = np.zeros((num_actors, len(seeds)), dtype=np.int64)
+        self.seed_staleness = np.array([0.]*len(seeds), dtype=np.float)
+
+        self.next_seed_index = 0 # Only used for sequential strategy
+
+    def seed_range(self):
+        return (int(min(self.seeds)), int(max(self.seeds)))
+
+    def _init_seed_index(self, seeds):
```
Comment (PaParaZz1) [inline ding/worker/level_replay/level_sampler.py:40]: merge these codes into `__init__`

File: ding/worker/level_replay/level_sampler.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+
+        # Update with a RolloutStorage object
+        if self.strategy == 'policy_entropy':
+            score_function = self._average_entropy
+        elif self.strategy == 'least_confidence':
+            score_function = self._average_least_confidence
+        elif self.strategy == 'min_margin':
+            score_function = self._average_min_margin
+        elif self.strategy == 'gae':
+            score_function = self._average_gae
```
Comment (PaParaZz1) [inline ding/worker/level_replay/level_sampler.py:56]: remove `average` prefix

File: ding/worker/level_replay/level_sampler.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+        rewards = kwargs['rewards']
+        value_preds = kwargs['value_preds']
+
+        max_t = len(rewards)
+        td_errors = (rewards[:-1] + value_preds[:max_t-1] - value_preds[1:max_t]).abs()
+
+        return td_errors.abs().mean().item()
+
+    @property
+    def requires_value_buffers(self):
```
Comment (PaParaZz1) [inline ding/worker/level_replay/level_sampler.py:134]: merge this property

File: ding/worker/collector/sample_serial_collector.py [hunk 8]
```diff
@@ -316,6 +316,135 @@ def collect(
         else:
             return return_data
 
+
+    def collect_plr(
```
Comment (PaParaZz1) [inline ding/worker/collector/sample_serial_collector.py:320]: merge this method into collect, add level_seeds default argument

File: ding/entry/serial_entry_plr.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+    # ==========
+    # Main loop
+    # ==========
+    # Learner's before_run hook.
+    learner.call_hook('before_run')
+    seeds = torch.zeros(collector_env_num)
+    for e in range(collector_env_num):
+        seed = level_sampler.sample('sequential')
+        seeds[e] = seed
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_plr.py:92]: `seeds =  [level_sampler.sample('sequential') for _ in range(collector_env_num)]`

File: ding/entry/serial_entry_plr.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    trans_seeds = seeds.tolist()
+    for e in range(collector_env_num):
+        trans_seeds[e] = int(trans_seeds[e])
+
+    collector_env.seed(trans_seeds)
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_plr.py:97]: set seed, then reset env at once

File: ding/entry/serial_entry_plr.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        # Evaluate policy performance
+        if evaluator.should_eval(learner.train_iter):
+            stop, reward = evaluator.eval(learner.save_checkpoint, learner.train_iter, collector.envstep)
+            if stop:
+                break
+        # Collect data by default config n_sample/n_episode
+        new_data = collector.collect_plr(train_iter=learner.train_iter, level_seeds= seeds, policy_kwargs=collect_kwargs)
+        # Learn policy from collected data
+        learner.train(new_data, collector.envstep)
+        if collector.envstep >= max_env_step or learner.train_iter >= max_train_iter:
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_plr.py:111]: move these two lines to the end of this while loop

File: ding/entry/serial_entry_plr.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+            break
+        train_data_2 = default_preprocess_learn(new_data, ignore_done=cfg.policy.learn.ignore_done, use_nstep=False)
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_plr.py:113]: rename to stacked_data

File: ding/worker/level_replay/level_sampler.py [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+        done = train_data['done'].reshape(num_actors, int(all_total_steps/num_actors)).transpose(0,1)
+        total_steps, num_actors = policy_logits.shape[:2]
+        num_decisions = len(policy_logits)
+
+        for actor_index in range(num_actors):
+            done_steps = done[:,actor_index].nonzero()[:total_steps,0]
+            start_t = 0
+
+            for t in done_steps:
+                if not start_t < total_steps: break
```
Comment (PaParaZz1) [inline ding/worker/level_replay/level_sampler.py:150]: break should occupy a solo line

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/entry/serial_entry_plr.py [hunk 14]
```diff
@@ -0,0 +1,126 @@
+import logging
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_plr.py:1]: add PLR into readme

**Comment by zerlinwang** [comment]
lastly, use yapf and flake8 to standardise code style, and check the failed unit_test

File: ding/worker/collector/sample_serial_collector.py [hunk 15]
```diff
@@ -195,7 +195,8 @@ def collect(
             n_sample: Optional[int] = None,
             train_iter: int = 0,
             drop_extra: bool = True,
-            policy_kwargs: Optional[dict] = None
+            policy_kwargs: Optional[dict] = None,
+            level_seeds: Optional[List] = None,
```
Comment (PaParaZz1) [inline ding/worker/collector/sample_serial_collector.py:199]: add comment for this argument

File: dizoo/procgen/entry/bigfish_plr_config.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+            clip_ratio=0.2,
+            aux_freq=1,
+        ),
+        collect=dict(n_sample=16384, ),
+        eval=dict(evaluator=dict(eval_freq=96, )),
+        other=dict(
+        ),
+    ),
+    level_replay=dict(
+        level_replay_strategy='min_margin',
```
Comment (PaParaZz1) [inline dizoo/procgen/entry/bigfish_plr_config.py:42]: maybe we can omit `level_replay` prefix

File: ding/entry/serial_entry_plr.py [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+    )
+
+    # ==========
+    # Main loop
+    # ==========
+    # Learner's before_run hook.
+    learner.call_hook('before_run')
+
+    seeds = [int(level_sampler.sample('sequential')) for _ in range(collector_env_num)]
+    level_seeds = torch.Tensor(seeds)
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_plr.py:101]: add comment for why transform into tensor

File: ding/data/level_replay/level_sampler.py [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+        episode_logits = kwargs['episode_logits']
+        return (1 - torch.exp(episode_logits.max(-1, keepdim=True)[0])).mean().item()
+
+    def _min_margin(self, **kwargs):
+        episode_logits = kwargs['episode_logits']
+        top2_confidence = torch.exp(episode_logits.topk(2, dim=-1)[0])
+        return 1 - (top2_confidence[:, 0] - top2_confidence[:, 1]).mean().item()
+
+    def _gae(self, **kwargs):
+        #returns = kwargs['returns']
```
Comment (PaParaZz1) [inline ding/data/level_replay/level_sampler.py:146]: remove commented code

File: ding/data/level_replay/level_sampler.py [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+            staleness_weights = staleness_weights * (1 - self.unseen_seed_weights)
+            z = np.sum(staleness_weights)
+            if z > 0:
+                staleness_weights /= z
+
+            weights = (1 - self.staleness_coef) * weights + self.staleness_coef * staleness_weights
+
+        return weights
+
+    def _score_transform(self, transform, temperature, scores):
```
Comment (PaParaZz1) [inline ding/data/level_replay/level_sampler.py:317]: add all python typing

File: ding/data/level_replay/level_sampler.py [hunk 20]
```diff
@@ -0,0 +1,327 @@
...
+        self._update_staleness(seed_idx)
+
+        return int(seed)
+
+    def sample(self, strategy=None):
+        if not strategy:
+            strategy = self.strategy
+
+        if strategy == 'random':
+            seed_idx = np.random.choice(range((len(self.seeds))))
```
Comment (PaParaZz1) [inline ding/data/level_replay/level_sampler.py:267]: range(len(self.seeds))

File: ding/data/level_replay/level_sampler.py [hunk 21]
```diff
@@ -0,0 +1,327 @@
...
+            seed = self.seeds[seed_idx]
+            return int(seed)
```
Comment (PaParaZz1) [inline ding/data/level_replay/level_sampler.py:269]: why int here

File: ding/data/level_replay/level_sampler.py [hunk 22]
```diff
@@ -0,0 +1,327 @@
...
+
+        num_unseen = (self.unseen_seed_weights > 0).sum()
+        proportion_seen = (len(self.seeds) - num_unseen) / len(self.seeds)
+
+        if self.replay_schedule == 'fixed':
+            if proportion_seen >= self.rho:
+                # Sample replay level with fixed prob = 1 - nu OR if all levels seen
+                if np.random.rand() > self.nu or not proportion_seen < 1.0:
```
Comment (PaParaZz1) [inline ding/data/level_replay/level_sampler.py:283]: np.isclose(proportion_seen, 1.0)?

File: ding/data/level_replay/level_sampler.py [hunk 23]
```diff
@@ -0,0 +1,327 @@
...
+                    score_function_kwargs['rewards'] = rewards[start_t:, actor_index]
+                    score_function_kwargs['value_preds'] = value[start_t:, actor_index]
+
+                score = score_function(**score_function_kwargs)
+                num_steps = len(episode_logits)
+                self._partial_update_seed_score(actor_index, seed_idx_t, score, num_steps)
+
+    def _update_staleness(self, selected_idx):
+        if self.staleness_coef > 0:
+            self.seed_staleness = self.seed_staleness + 1
```
Comment (PaParaZz1) [inline ding/data/level_replay/level_sampler.py:237]: += 1

File: ding/data/level_replay/level_sampler.py [hunk 24]
```diff
@@ -0,0 +1,327 @@
...
+
+                if self.strategy in ['gae', 'value_l1', 'one_step_td_error']:
+                    rewards = train_data['reward'].reshape(num_actors,
+                                                           int(all_total_steps / num_actors)).transpose(0, 1)
+                    adv = train_data['adv'].reshape(num_actors, int(all_total_steps / num_actors)).transpose(0, 1)
+                    value = train_data['value'].reshape(num_actors, int(all_total_steps / num_actors)).transpose(0, 1)
+                    #score_function_kwargs['returns'] = rollouts.returns[start_t:t,actor_index]
+                    score_function_kwargs['adv'] = adv[start_t:t, actor_index]
+                    score_function_kwargs['rewards'] = rewards[start_t:t, actor_index]
+                    score_function_kwargs['value_preds'] = value[start_t:t, actor_index]
```
Comment (PaParaZz1) [inline ding/data/level_replay/level_sampler.py:206]: use our rename method rather than something like `value_preds`

File: ding/data/level_replay/level_sampler.py [hunk 25]
```diff
@@ -0,0 +1,327 @@
...
+        return advantages.mean().item()
+
+    def _value_l1(self, **kwargs):
+        #returns = kwargs['returns']
+        #value_preds = kwargs['value_preds']
+
+        #advantages = returns - value_preds
+        advantages = kwargs['adv']
+
+        return advantages.abs().mean().item()
```
Comment (PaParaZz1) [inline ding/data/level_replay/level_sampler.py:161]: why abs here, add some comments about motivation

File: ding/data/level_replay/level_sampler.py [hunk 26]
```diff
@@ -0,0 +1,327 @@
...
+        elif self.strategy == 'min_margin':
+            score_function = self._min_margin
+        elif self.strategy == 'gae':
+            score_function = self._gae
+        elif self.strategy == 'value_l1':
+            score_function = self._value_l1
+        elif self.strategy == 'one_step_td_error':
+            score_function = self._one_step_td_error
+        else:
+            raise ValueError(f'Unsupported strategy, {self.strategy}')
```
Comment (PaParaZz1) [inline ding/data/level_replay/level_sampler.py:94]: use our format, like `'Not supported strategy: {}'.format(self.strategy)`

File: ding/data/level_replay/level_sampler.py [hunk 27]
```diff
@@ -0,0 +1,327 @@
...
+
+        self.unseen_seed_weights = np.ones(len(seeds))
+        self.seed_scores = np.zeros(len(seeds))
+        self.partial_seed_scores = np.zeros((num_actors, len(seeds)), dtype=np.float)
+        self.partial_seed_steps = np.zeros((num_actors, len(seeds)), dtype=np.int64)
+        self.seed_staleness = np.zeros(len(seeds))
+
+        self.next_seed_index = 0  # Only used for sequential strategy
+
+    def seed_range(self):
```
Comment (PaParaZz1) [inline ding/data/level_replay/level_sampler.py:73]: where is this method used

File: ding/data/level_replay/level_sampler.py [hunk 28]
```diff
@@ -0,0 +1,327 @@
...
+    def __init__(
+        self,
+        seeds: Optional[List[int]],
+        obs_space: Union[int, SequenceType],
+        action_space: int,
+        num_actors: int = 1,
+        strategy: Optional[str] = 'random',
+        replay_schedule: Optional[str] = 'fixed',
+        score_transform: Optional[str] = 'power',
+        temperature: float = 1.0,
```
Comment (PaParaZz1) [inline ding/data/level_replay/level_sampler.py:34]: add this arguments to config

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/data/level_replay/level_sampler.py (hunk 21 above)
Comment (Weiyuhong-1998) [reply ding/data/level_replay/level_sampler.py:269]: seed must be int data

**Review by PaParaZz1 — APPROVED** [review]
